src/components/workspace/editorMode/source.js

```diff
@@ -0,0 +1,831 @@
+    
+    var oop = acequire("../lib/oop");
+    var TextMode = acequire("./text").Mode;
+    var JavaScriptHighlightRules = acequire("./javascript_highlight_rules").JavaScriptHighlightRules;
+    var MatchingBraceOutdent = acequire("./matching_brace_outdent").MatchingBraceOutdent;
+    var WorkerClient = acequire("../worker/worker_client").WorkerClient;
+    var CstyleBehaviour = acequire("./behaviour/cstyle").CstyleBehaviour;
+    var CStyleFoldMode = acequire("./folding/cstyle").FoldMode;
+    
+    var Mode = function() {
+        this.HighlightRules = JavaScriptHighlightRules;
+        
+        this.$outdent = new MatchingBraceOutdent();
+        this.$behaviour = new CstyleBehaviour();
+        this.foldingRules = new CStyleFoldMode();
+    };
+    oop.inherits(Mode, TextMode);
+    
+    (function() {
+    
+        this.lineCommentStart = "//";
+        this.blockComment = {start: "/*", end: "*/"};
+        this.$quotes = {'"': '"', "'": "'", "`": "`"};
+    
+        this.getNextLineIndent = function(state, line, tab) {
+            var indent = this.$getIndent(line);
+    
+            var tokenizedLine = this.getTokenizer().getLineTokens(line, state);
+            var tokens = tokenizedLine.tokens;
+            var endState = tokenizedLine.state;
+            var match = null;
+
+            if (tokens.length && tokens[tokens.length-1].type === "comment") {
+                return indent;
+            }
+    
+            if (state === "start" || state === "no_regex") {
+                match = line.match(/^.*(?:\bcase\b.*:|[\{\(\[])\s*$/);
+                if (match) {
+                    indent += tab;
+                }
+            } else if (state === "doc-start") {
+                if (endState === "start" || endState === "no_regex") {
+                    return "";
+                }
+                match = line.match(/^\s*(\/?)\*/);
+                if (match) {
+                    if (match[1]) {
+                        indent += " ";
+                    }
+                    indent += "* ";
+                }
+            }
+    
+            return indent;
+        };
+    
+        this.checkOutdent = function(state, line, input) {
+            return this.$outdent.checkOutdent(line, input);
+        };
+    
+        this.autoOutdent = function(state, doc, row) {
+            this.$outdent.autoOutdent(doc, row);
+        };
+    
+        this.createWorker = function(session) {
+            var worker = new WorkerClient(["ace"], require("../editorWorker/source.js"), "JavaScriptWorker");
+            worker.attachToDocument(session.getDocument());
+    
+            worker.on("annotate", function(results) {
+                session.setAnnotations(results.data);
+            });
+    
+            worker.on("terminate", function() {
+                session.clearAnnotations();
+            });
+    
+            return worker;
+        };
+    
+        this.$id = "ace/mode/javascript";
+    }).call(Mode.prototype);
```

Comment:
Again, I am confused by this notation, where an anonymous function is immediately run

Reply:
I tried separate the function out and refactor it to ES2015 class syntax. But the code crashed after i did that.

Reply:
Ok, no problem